composeApp/src/commonMain/kotlin/org/ooni/probe/data/repositories/PreferenceRepository.kt

```diff
@@ -104,6 +139,34 @@ class PreferenceRepository(
         }
     }
 
+    suspend fun getProxyString(): String {
+        val proxyHost =
+            getValueByKey(SettingsKey.PROXY_HOSTNAME).firstOrNull() as String? ?: return ""
+        val proxyPort = getValueByKey(SettingsKey.PROXY_PORT).firstOrNull() as Int? ?: return ""
+
+        when (
+            val proxyProtocol =
+                getValueByKey(SettingsKey.PROXY_PROTOCOL).firstOrNull() as String?
+        ) {
+            ProxyProtocol.NONE.protocol -> return ""
+            ProxyProtocol.PSIPHON.protocol -> return "psiphon://"
+            ProxyProtocol.SOCKS5.protocol, ProxyProtocol.HTTP.protocol, ProxyProtocol.HTTPS.protocol -> {
+                val formattedHost = if (isIPv6(proxyHost)) {
+                    "[$proxyHost]"
+                } else {
+                    proxyHost
+                }
+                return "$proxyProtocol://$formattedHost:$proxyPort/"
+            }
+
+            else -> return ""
+        }
+    }
```

Comment:
I feel all of this is more than the PreferencesRepository should know about the Proxy logic. At most it could receive a ProxySettings object and know how to split that into individual preferences, and the other way around.

But generating a proxy string could happen inside a ProxySettings object. The validation could happen on ProxySettings, in the ViewModel, or the domain.